actioncable/lib/action_cable/connection/client_socket.rb

```diff
@@ -0,0 +1,177 @@
+require 'websocket/driver'
+
+module ActionCable
+  module Connection
+    #--
+    # This class is heavily based on faye-websocket-ruby
+    #
+    # Copyright (c) 2010-2015 James Coglan
+    class ClientSocket
+      def self.determine_url(env)
+        scheme = secure_request?(env) ? 'wss:' : 'ws:'
+        "#{ scheme }//#{ env['HTTP_HOST'] }#{ env['REQUEST_URI'] }"
+      end
+
+      def self.secure_request?(env)
+        return true if env['HTTPS'] == 'on'
+        return true if env['HTTP_X_FORWARDED_SSL'] == 'on'
+        return true if env['HTTP_X_FORWARDED_SCHEME'] == 'https'
+        return true if env['HTTP_X_FORWARDED_PROTO'] == 'https'
+        return true if env['rack.url_scheme'] == 'https'
+
+        return false
+      end
+
+      attr_reader :env
+
+      def initialize(env, event_target, stream_event_loop, protocols = nil, options = {})
+        @env = env
+        @event_target = event_target
+        @stream_event_loop = stream_event_loop
+
+        @url = ClientSocket.determine_url(@env)
+
+        @driver = @driver_started = nil
+
+        @ready_state = CONNECTING
+        super()
+        ::WebSocket::Driver.validate_options(options, [:headers, :extensions, :max_length, :ping, :tls])
+
+        @driver = ::WebSocket::Driver.rack(self, :max_length => options[:max_length], :protocols => protocols)
+
+        if headers = options[:headers]
+          headers.each { |name, value| @driver.set_header(name, value) }
+        end
+
+        [*options[:extensions]].each do |extension|
```

Comment:
why use two different array handling styles between `:headers` and `:extensions`?